### PR TITLE
Hardsuits, EVA suits, Firesuits, and others now protect your feet from Glass Shards

### PR DESCRIPTION
**Pull request by Plykiya**
## About the PR

Added the ShoeRequiredStepTriggerImmune comp to anything I could find with shoes in the sprite.
- All hardsuits
- All EVA suits
- Both bomb suits
- Both fire suits
- Rad suits
- Chitinous armor
- Bone armor
- ironclad II armor
- heavy armor
- Ian suit
- Chicken suit
- Monkey suit

## Why / Balance

MY FEET HURT

## Media

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

**Breaking Changes**
- Removed ShoeRequiredStepTriggerImmune tag and turned it into a comp called ShoeRequiredStepTriggerImmuneComp

**Changelog**

:cl:
- tweak: Hardsuits, EVA suits, firesuits and other things now protect your feet from dangerous glass shards.

### CODE REVIEW

**Comment by Plykiya** [comment]
tested it with the syndicate EVA and other EVA suits, good
tested it with hardsuits like the captain and engi suits, good

**Comment by Plykiya** [comment]
alternatively I can add a "hasShoes" tag or something and rely on that so we don't end up having to manually add in more to the condition check

**Comment by FungiFellow** [comment]
Are Firesuits included?

**Comment by Plykiya** [comment]
> Are Firesuits included?

ah... good point. I think I will just add that shoe tag...

**Comment by Plykiya** [comment]
I just realized what I need to do, one second.

**Comment by Plykiya** [comment]
Alright now it's a yml only change. I added the immune tag to anything armor-wise I could find that had shoes.

**Comment by Plykiya** [comment]
The negative addition is just me altering the atmos fire suit indentation to match the rest of the file.

**Comment by Plykiya** [comment]
the only thing this PR doesn't address yet is the fact that Dionas and Dionukies also benefit from the hardsuit changes since it only checks for outerClothing. let me see what I can do to make Dionas suffer

**Comment by Dutch-VanDerLinde** [comment]
> the only thing this PR doesn't address yet is the fact that Dionas and Dionukies also benefit from the hardsuit changes since it only checks for outerClothing. let me see what I can do to make Dionas suffer

??? whats the point of this change if dionas dont benefit from it

**Comment by Plykiya** [comment]
> > the only thing this PR doesn't address yet is the fact that Dionas and Dionukies also benefit from the hardsuit changes since it only checks for outerClothing. let me see what I can do to make Dionas suffer
> 
> ??? whats the point of this change if dionas dont benefit from it

the CE getting their shoes stolen and breaking their fucking toes on glass shards despite having a chief engineering hardsuit

**Comment by Plykiya** [comment]
there are arguments to be made about Dionas benefitting from the shoes because the shoe sprites still show up when they wear it

I say that this is an oversight and maybe displacement maps can vanish them from existence

**Comment by Dutch-VanDerLinde** [comment]
c'mon man they already cant wear magboots or noslips why make them able to be defeated by a glass shard :(

**Comment by HelpTheSteak** [comment]
> there are arguments to be made about Dionas benefitting from the shoes because the shoe sprites still show up when they wear it
> 
> I say that's just a limitation of the current clothing implementations

But wouldn't that mean hardsuits on dionas would also be useless due to the gaping hole on the bottom

**Comment by Plykiya** [comment]
> > there are arguments to be made about Dionas benefitting from the shoes because the shoe sprites still show up when they wear it
> > I say that's just a limitation of the current clothing implementations
> 
> But wouldn't that mean hardsuits on dionas would also be useless due to the gaping hole on the bottom

I'm all for buffing Dionas if a maintainer wants to remove the change, I'm just going with the same logic as lizard tails coming out of their spacesuit not compromising them either

**Comment by HelpTheSteak** [comment]
> > > there are arguments to be made about Dionas benefitting from the shoes because the shoe sprites still show up when they wear it
> > > I say that's just a limitation of the current clothing implementations
> > 
> > 
> > But wouldn't that mean hardsuits on dionas would also be useless due to the gaping hole on the bottom
> 
> I'm all for buffing Dionas if a maintainer wants to remove the change, I'm just going with the same logic as lizard tails coming out of their spacesuit not compromising them either...
> 
> the holes are... magically fine

You know what I'm all for hardsuits for dionas and lizards being made useless if it makes slimes a little better in comparison

**Comment by Plykiya** [comment]
> > > > there are arguments to be made about Dionas benefitting from the shoes because the shoe sprites still show up when they wear it
> > > > I say that's just a limitation of the current clothing implementations
> > > 
> > > 
> > > But wouldn't that mean hardsuits on dionas would also be useless due to the gaping hole on the bottom
> > 
> > 
> > I'm all for buffing Dionas if a maintainer wants to remove the change, I'm just going with the same logic as lizard tails coming out of their spacesuit not compromising them either...
> > the holes are... magically fine
> 
> You know what I'm all for hardsuits for dionas and lizards being made useless if it makes slimes a little better in comparison

the most based take imaginable
dionas and lizards becoming more unique by being unable to safely exist in a spaced environment due to their biology...

**Review by EmoGarbage404 — CHANGES_REQUESTED** [review]
this code would be cleanest if `ShoesRequiredStepTriggerImmune` was just converted to a component and then all worn clothing slots (excluding pockets) were checked for instances of this component. Needing to special case in every clothing slot that can feasibly cover your feet is really bad for expanding and this code is getting duplicated with these changes enough as it is.

**Comment by Plykiya** [comment]
> this code would be cleanest if `ShoesRequiredStepTriggerImmune` was just converted to a component and then all worn clothing slots (excluding pockets) were checked for instances of this component. Needing to special case in every clothing slot that can feasibly cover your feet is really bad for expanding and this code is getting duplicated with these changes enough as it is.

okay, how's this

File: Content.Shared/StepTrigger/Systems/ShoesRequiredStepTriggerSystem.cs [hunk 1]
```diff
@@ -19,18 +18,33 @@ public override void Initialize()
...
+        if (!_inventory.HasSlot(args.Tripper, "shoes", inventory))
+            return;
+
+        // go through all equipped items, checks if item is equipped in shoe slot or has shoe immmune component
+        if (_inventory.TryGetContainerSlotEnumerator(args.Tripper, out var containerSlotEnumerator))
         {
-            args.Cancelled = true;
+            while (containerSlotEnumerator.NextItem(out var item, out var slot))
+            {
+                if (slot.Name == "shoes" || HasComp<ShoesRequiredStepTriggerImmuneComponent>(item))
```
Comment (EmoGarbage404) [inline Content.Shared/StepTrigger/Systems/ShoesRequiredStepTriggerSystem.cs:42]: just put ShoesRequiredStepTriggerComponent on baseShoes

File: Content.Shared/StepTrigger/Systems/ShoesRequiredStepTriggerSystem.cs [hunk 2]
```diff
@@ -19,18 +18,33 @@ public override void Initialize()
...
+        // early exit if no shoe slot exists at all, dionas suffer
+        // remove to give dionas their freedom from glass
+        if (!_inventory.HasSlot(args.Tripper, "shoes", inventory))
+            return;
+
+        // go through all equipped items, checks if item is equipped in shoe slot or has shoe immmune component
+        if (_inventory.TryGetContainerSlotEnumerator(args.Tripper, out var containerSlotEnumerator))
         {
-            args.Cancelled = true;
+            while (containerSlotEnumerator.NextItem(out var item, out var slot))
```
Comment (EmoGarbage404) [inline Content.Shared/StepTrigger/Systems/ShoesRequiredStepTriggerSystem.cs:40]: you need to check that these items are actually in their appropriate item slots and not in pockets or something. I think this is done with relay events most of the time, so you should look into this.

File: Content.Shared/StepTrigger/Systems/ShoesRequiredStepTriggerSystem.cs [hunk 3]
```diff
@@ -19,18 +18,33 @@ public override void Initialize()
...
 
+        // if not, checks the inventory
         if (!TryComp<InventoryComponent>(args.Tripper, out var inventory))
             return;
 
-        if (_inventory.TryGetSlotEntity(args.Tripper, "shoes", out _, inventory))
+        // early exit if no shoe slot exists at all, dionas suffer
+        // remove to give dionas their freedom from glass
+        if (!_inventory.HasSlot(args.Tripper, "shoes", inventory))
+            return;
```
Comment (EmoGarbage404) [inline Content.Shared/StepTrigger/Systems/ShoesRequiredStepTriggerSystem.cs:35]: this is horribly janky and really just confusing logic. Hardsuits protect from stepping on glass unless you don't have shoe slots in which case they suddenly stop doing that? 

just remove the whole chunk IMO

File: Content.Shared/StepTrigger/Components/ShoesRequiredStepTriggerImmuneComponent.cs [hunk 4]
```diff
@@ -0,0 +1,11 @@
+using Robust.Shared.GameStates;
+
+namespace Content.Shared.StepTrigger.Components;
+
+/// <summary>
+/// This is used for cancelling step trigger events if the user is wearing shoes, such as for glass shards.
+/// </summary>
+[RegisterComponent, NetworkedComponent]
```
Comment (EmoGarbage404) [inline Content.Shared/StepTrigger/Components/ShoesRequiredStepTriggerImmuneComponent.cs:8]: add Access for ShoeRequiredStepTriggerSystem

**Review by EmoGarbage404 — CHANGES_REQUESTED** [review]

File: Content.Shared/StepTrigger/Systems/ShoesRequiredStepTriggerSystem.cs (hunk 2 above)
Comment (Plykiya) [reply Content.Shared/StepTrigger/Systems/ShoesRequiredStepTriggerSystem.cs:31]: okay, how about this...

File: Content.Shared/StepTrigger/Systems/ShoesRequiredStepTriggerSystem.cs (hunk 3 above)
Comment (Plykiya) [reply Content.Shared/StepTrigger/Systems/ShoesRequiredStepTriggerSystem.cs]: Dionas shall receive their freedom....

File: Content.Shared/StepTrigger/Systems/ShoesRequiredStepTriggerSystem.cs (hunk 2 above)
Comment (Plykiya) [reply Content.Shared/StepTrigger/Systems/ShoesRequiredStepTriggerSystem.cs:31]: I'll look through the relay events now

Comment (Plykiya) [reply Content.Shared/StepTrigger/Systems/ShoesRequiredStepTriggerSystem.cs:31]: If you want me to use relay events then I'll likely have to check at the time anything is equipped and check if anything has the comp and save that status. That way the player's entire inventory isn't being checked every time they slip, but instead only when their equipped item changes. Will work on that...

Comment (Plykiya) [reply Content.Shared/StepTrigger/Systems/ShoesRequiredStepTriggerSystem.cs:31]: Currently I re-implemented it to just check every slot besides pockets...

**Comment by Plykiya** [comment]
I.... I don't really know how to tie this into any kind of inventory equip event, considering that the moment you step on a piece of glass or a D4 isn't really tied to any kind "inventory equip" event

File: Content.Shared/StepTrigger/Systems/ShoesRequiredStepTriggerSystem.cs (hunk 2 above)
Comment (metalgearsloth) [reply Content.Shared/StepTrigger/Systems/ShoesRequiredStepTriggerSystem.cs:31]: I think we have relay events for specific slots now?

**Pull request merged**